crates/subspace-farmer/src/bin/subspace-farmer/commands/farm.rs

```diff
@@ -192,10 +214,22 @@ pub(crate) async fn farm_multi_disk(
         .map(|single_disk_farm| single_disk_farm.wait())
         .collect::<FuturesUnordered<_>>();
 
-    while let Some(result) = single_disk_farms_stream.next().await {
-        result?;
-
-        info!("Farm exited successfully");
+    loop {
+        futures::select! {
+            maybe_result = single_disk_farms_stream.next().fuse() => {
+                if let Some(result) = maybe_result {
+                    result?;
+                    info!("Farm exited successfully");
+                } else {
+                    break
+                }
+            },
+            result = received_signal => {
+                if result.is_ok() {
+                    break
+                }
+            }
+        }
     }
```

Comment:
I don't think you need a loop here, just race against received signal against `single_disk_farms_stream.next().fuse()` and you'll achieve the same result.